pkg/util/validation/validate.go

```diff
@@ -146,6 +164,44 @@ func ValidateLabels(cfg LabelValidationConfig, userID string, ls []client.LabelA
 	return nil
 }
 
+// MetadataValidationConfig helps with getting required config to validate metadata.
+type MetadataValidationConfig interface {
+	EnforceMetadataMetricName(userID string) bool
+	MaxMetadataLength(userID string) int
+}
+
+// ValidateMetadata returns an err if a metric metadata is invalid.
+func ValidateMetadata(cfg MetadataValidationConfig, userID string, metadata *client.MetricMetadata) error {
+	if cfg.EnforceMetadataMetricName(userID) && metadata.MetricName == "" {
+		return httpgrpc.Errorf(http.StatusBadRequest, errMetadataMissingMetricName)
+	}
+
+	maxMetadataValueLength := cfg.MaxMetadataLength(userID)
+	if len(metadata.MetricName) > maxMetadataValueLength {
+		DiscardedMetadata.WithLabelValues(metricNameTooLong, userID).Inc()
+		return httpgrpc.Errorf(http.StatusBadRequest, errMetadataMetricNameTooLong, metadata.MetricName)
+	}
```

Comment:
I have changed it, but I think it reads a bit too convoluted e.g. `metadata 'METRIC_NAME' value too long: \"go_goroutines_and_routines_and_routines\" metric \"go_goroutines_and_routines_and_routines\"` as we need to append the metric name at the end for the non-metric name failures.

Happy to keep it as is if you think it's OK though.

Reply:
Maybe we can improve the error message? Something like:

```
errMetadataTooLong = "%s metadata value too long for the metric: %.200q value %.200q"
```

In the case of the metric name it will be displayed twice, but doesn't look a big deal to me (as far as it's very clear what the error is).

Reply:
I'll leave it as is then. I'd prefer to keep the `metric X` format, as is the same we use or samples - at least it's consistent.